pennylane/vqe/vqe.py

```diff
@@ -468,6 +470,20 @@ def __init__(
         """QNodeCollection: The QNodes to be evaluated. Each QNode corresponds to the expectation
         value of each observable term after applying the circuit ansatz."""
 
+        wires = device.wires.tolist()
+
+        @old_qnode(device, interface=interface, diff_method=diff_method, **kwargs)
+        def qnode_for_metric_tensor_in_tape_mode(*qnode_args, _wires=wires, **qnode_kwargs):
+            """The metric tensor cannot currently be calculated in tape-mode QNodes. As a
+            short-term fix for VQECost, we create a non-tape mode QNode just for calculation of the
+            metric tensor. In doing so, we reintroduce the same restrictions of the old QNode but
+            allow users to access new functionality such as measurement grouping and batch
+            execution of the gradient."""
+            ansatz(*qnode_args, wires=_wires, **qnode_kwargs)
+            return old_expval(qml.PauliZ(0))
+
+        self._qnode_for_metric_tensor_in_tape_mode = qnode_for_metric_tensor_in_tape_mode
+
```

Comment:
An alternative could be to disable and then re-enable tape mode. However this does not seem to work:

```python
import pennylane as qml

dev = qml.device("default.qubit", wires=2)

qml.enable_tape()

@qml.qnode(dev)
def f(x):
    qml.RX(x, wires=0)
    qml.RY(0.4, wires=1)
    qml.CNOT(wires=[0, 1])
    return qml.expval(qml.PauliZ(0))

qml.disable_tape()

@qml.qnode(dev)
def f2(x):
    qml.RX(x, wires=0)
    qml.RY(0.4, wires=1)
    qml.CNOT(wires=[0, 1])
    return qml.expval(qml.PauliZ(0))

qml.enable_tape()

print(f(0.4))
print(f2(0.4))
```
The second line gives:
`QuantumFunctionError: A quantum function must return either a single measured observable or a nonempty sequence of measured observables.`